### PR TITLE
feat: qv impact stream refactor

### PR DESCRIPTION
**Pull request by 0xOneTony**
QVImpactStream strategy refactor leveraging the new strategies and tools.

### CODE REVIEW

File: contracts/strategies/QVImpactStream.sol [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+// ⠀⠀⠀⠀⠀⠀⠀⣴⣿⣿⣿⣿⣿⣿⡿⠃⠀⠙⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⣿⣿⠁⠀⠀⠀⢻⣿⣿⣿⣧⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⠀⠀⣠⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣶⡀⠀⠀⠀⠀
+// ⠀⠀⠀⠀⠀⢀⣾⣿⣿⣿⣿⣿⣿⡿⠁⠀⠀⠀⠘⣿⣿⣿⣿⣿⡿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣾⣿⣿⣿⠃⠀⠀⠀⠀⠈⢿⣿⣿⣿⣆⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⠀⣰⣿⣿⣿⡿⠋⠁⠀⠀⠈⠘⠹⣿⣿⣿⣿⣆⠀⠀⠀
+// ⠀⠀⠀⠀⢀⣾⣿⣿⣿⣿⣿⣿⡿⠀⠀⠀⠀⠀⠀⠈⢿⣿⣿⣿⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣾⣿⣿⣿⠏⠀⠀⠀⠀⠀⠀⠘⣿⣿⣿⣿⡄⠀⠀⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⢰⣿⣿⣿⣿⠁⠀⠀⠀⠀⠀⠀⠀⠘⣿⣿⣿⣿⡀⠀⠀
+// ⠀⠀⠀⢠⣿⣿⣿⣿⣿⣿⣿⣟⠀⡀⢀⠀⡀⢀⠀⡀⢈⢿⡟⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣼⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡄⠀⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⣿⣿⣿⡇⠀⠀
+// ⠀⠀⣠⣿⣿⣿⣿⣿⣿⡿⠋⢻⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣶⣄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣸⣿⣿⣿⡿⢿⠿⠿⠿⠿⠿⠿⠿⠿⠿⢿⣿⣿⣿⣷⡀⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⠸⣿⣿⣿⣷⡀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⣿⣿⠂⠀⠀
+// ⠀⠀⠙⠛⠿⠻⠻⠛⠉⠀⠀⠈⢿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣰⣿⣿⣿⣿⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢿⣿⣿⣿⣧⠀⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⠀⢻⣿⣿⣿⣷⣀⢀⠀⠀⠀⡀⣰⣾⣿⣿⣿⠏⠀⠀⠀
+// ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠛⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⢰⣿⣿⣿⣿⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠘⣿⣿⣿⣿⣧⠀⠀⢸⣿⣿⣿⣗⠀⠀⠀⢸⣿⣿⣿⡯⠀⠀⠀⠀⠹⢿⣿⣿⣿⣿⣾⣾⣷⣿⣿⣿⣿⡿⠋⠀⠀⠀⠀
+// ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠙⠙⠋⠛⠙⠋⠛⠙⠋⠛⠙⠋⠃⠀⠀⠀⠀⠀⠀⠀⠀⠠⠿⠻⠟⠿⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⠟⠿⠟⠿⠆⠀⠸⠿⠿⠟⠯⠀⠀⠀⠸⠿⠿⠿⠏⠀⠀⠀⠀⠀⠈⠉⠻⠻⡿⣿⢿⡿⡿⠿⠛⠁⠀⠀⠀⠀⠀⠀
+//                    allo.gitcoin.co
+contract QVImpactStream is QVSimple, Multicall {
```
Comment (ilpepepig) [inline contracts/strategies/QVImpactStream.sol:27]: Inheriting from [DonationVotingOffchain](https://github.com/allo-protocol/allo-v2/blob/v2.1/contracts/strategies/DonationVotingOffchain.sol) might be a better fit than [QVSimple](https://github.com/allo-protocol/allo-v2/blob/f01c3ff847db25e0d80136d3a0eda93b5d840bbe/contracts/strategies/QVSimple.sol)

Comment (0xOneTony) [reply contracts/strategies/QVImpactStream.sol:27]: I see your point with the payout and claim, but since this is a QV strategy I don't think it makes sense to inherit from basically QF, I would still need to re-write the allocation logic. I think it makes more sense if we have an extension for `OffchainPayout` or something like that, that includes these `setPayout`, `claim` etc

File: contracts/strategies/QVSimple.sol [hunk 2]
```diff
@@ -228,6 +224,24 @@ contract QVSimple is CoreBaseStrategy, RecipientsExtension {
...
+
+    /// @notice Remove allocator
+    /// @dev Only the pool manager(s) can call this function and emits an `AllocatorRemoved` event
+    /// @param _allocator The allocator address
+    function _removeAllocator(address _allocator) internal virtual {
+        allowedAllocators[_allocator] = false;
+
+        emit AllocatorRemoved(_allocator, msg.sender);
+    }
+
```
Comment (0xAustrian) [inline contracts/strategies/QVSimple.sol:244]: This sequence is repeated in a lot of strats, initially I thought about an `AllocatorsExtension`. It should include `add/remove` functions along with the allocation time variables. wdty?

File: contracts/strategies/QVImpactStream.sol [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    /// ======================
+    /// ======= Storage ======
+    /// ======================
+
+    /// @notice Returns the amount to pay to the recipient
+    /// @dev recipientId => payouts
+    mapping(address => uint256) public payouts;
+
+    bool public payoutSet;
```
Comment (0xAustrian) [inline contracts/strategies/QVImpactStream.sol:55]: Missing natspec

File: contracts/strategies/QVImpactStream.sol [hunk 4]
```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.19;
+
+// External Libraries
+import {Multicall} from "openzeppelin-contracts/contracts/utils/Multicall.sol";
```
Comment (0xAustrian) [inline contracts/strategies/QVImpactStream.sol:5]: Why do we need this library for? I think if we have batch functions we dont need it.

File: contracts/strategies/QVImpactStream.sol [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+import {IERC20Upgradeable} from "openzeppelin-contracts-upgradeable/contracts/token/ERC20/IERC20Upgradeable.sol";
```
Comment (0xAustrian) [inline contracts/strategies/QVImpactStream.sol:6]: You can use normal `IERC20` here. It will make no difference but it's correct.

File: contracts/strategies/QVImpactStream.sol [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        uint256 totalAmount;
+
+        uint256 length = _payouts.length;
+        for (uint256 i = 0; i < length;) {
+            Payout memory payout = _payouts[i];
+            uint256 amount = payout.amount;
+            address recipientId = payout.recipientId;
+
+            if (amount == 0 || _getRecipientStatus(recipientId) != Status.Accepted) {
+                revert RECIPIENT_ERROR(payout.recipientId);
```
Comment (0xAustrian) [inline contracts/strategies/QVImpactStream.sol:123]:
```suggestion
                revert RECIPIENT_ERROR(recipientId);
```

File: contracts/strategies/QVImpactStream.sol [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    /// @dev 'msg.sender' must be pool manager
+    /// @param _token The token to transfer
+    /// @param _recipient The recipient
+    function recoverFunds(address _token, address _recipient) external onlyPoolManager(msg.sender) {
+        // Get the amount of the token to transfer, which is always the entire balance of the contract address
+        uint256 amount = _token == NATIVE ? address(this).balance : IERC20Upgradeable(_token).balanceOf(address(this));
+
+        // Transfer the amount to the recipient (pool owner)
+        _transferAmount(_token, _recipient, amount);
+    }
```
Comment (0xAustrian) [inline contracts/strategies/QVImpactStream.sol:204]: I was wondering if the `withdraw` function can achieve the same goal.

File: test/foundry/strategies/QVImpactStream.t.sol [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+        /// since the recipient is not registered, his status should be NONE
+        /// so setting payouts should fail
+        vm.prank(poolManager);
+        qvImpactStream.setPayouts(payouts);
+    }
+
+    function test_SetPayoutsRevertWhen_TotalPayoutIsGreaterThanPoolAmount() external {
+        // it should revert
+        vm.skip(true);
```
Comment (0xAustrian) [inline test/foundry/strategies/QVImpactStream.t.sol:141]: Why those tests skipped?

File: contracts/strategies/QVImpactStream.sol (hunk 7 above)
Comment (0xOneTony) [reply contracts/strategies/QVImpactStream.sol:204]: fair, yes we can use withdraw instead

File: test/foundry/strategies/QVImpactStream.t.sol (hunk 8 above)
Comment (0xOneTony) [reply test/foundry/strategies/QVImpactStream.t.sol:141]: Not really sure what would be the best way to mock `_getRecipientStatus` ?

**Review by 0xAustrian — APPROVED** [review]

**Comment by ilpepepig** [comment]
The implementation looks good, but the strategy idea looks weird to me. Why would anyone use QVImpactStream instead of QVSimple which automatically calculates the payouts? Is there any reason the pool manager would deviate from the QV calculated payouts?

**Review by thelostone-mc — APPROVED** [review]

**Comment by thelostone-mc** [comment]
> Why would anyone use QVImpactStream instead of QVSimple which automatically calculates the payouts

TLDR: it was a customer request when we built out the strategy for https://www.impact.stream/
No technical reasoning for doing this

**Review by ilpepepig — APPROVED** [review]
